From 3c947725902977434361f682b1e93b964dd206e9 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Wed, 27 Jun 2018 16:29:51 +0800
Subject: [PATCH 4534/5242] MLK-19413-23 gpu: imx: dpu: Add
 constframe_framedimenstions_copy_prim() helper
 support

commit  0c7f97668a1e516d014d327826620bf8f8190f47 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds constframe_framedimenstions_copy_prim() helper support
so that callers may may copy frame dimensions from a primary constframe
to the relevant secondary constframe.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-constframe.c |   37 ++++++++++++++++++++++++++++++++++
 include/video/dpu.h                  |    1 +
 2 files changed, 38 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-constframe.c b/drivers/gpu/imx/dpu/dpu-constframe.c
index e25bd7d..a7c2719 100644
--- a/drivers/gpu/imx/dpu/dpu-constframe.c
+++ b/drivers/gpu/imx/dpu/dpu-constframe.c
@@ -53,6 +53,11 @@ struct dpu_constframe {
 	shadow_load_req_t shdlreq;
 };
 
+static inline u32 dpu_cf_read(struct dpu_constframe *cf, unsigned int offset)
+{
+	return readl(cf->base + offset);
+}
+
 static inline void dpu_cf_write(struct dpu_constframe *cf, u32 value,
 				unsigned int offset)
 {
@@ -84,6 +89,38 @@ void constframe_framedimensions(struct dpu_constframe *cf, unsigned int w,
 }
 EXPORT_SYMBOL_GPL(constframe_framedimensions);
 
+void constframe_framedimensions_copy_prim(struct dpu_constframe *cf)
+{
+	struct dpu_constframe *prim_cf = NULL;
+	unsigned int prim_id;
+	int i;
+	u32 val;
+
+	if (cf->id != 0 && cf->id != 1) {
+		dev_warn(cf->dpu->dev, "ConstFrame%d is not a secondary one\n",
+								cf->id);
+		return;
+	}
+
+	prim_id = cf->id + 4;
+
+	for (i = 0; i < ARRAY_SIZE(cf_ids); i++)
+		if (cf_ids[i] == prim_id)
+			prim_cf = cf->dpu->cf_priv[i];
+
+	if (!prim_cf) {
+		dev_warn(cf->dpu->dev, "cannot find ConstFrame%d's primary peer\n",
+								cf->id);
+		return;
+	}
+
+	mutex_lock(&cf->mutex);
+	val = dpu_cf_read(prim_cf, FRAMEDIMENSIONS);
+	dpu_cf_write(cf, val, FRAMEDIMENSIONS);
+	mutex_unlock(&cf->mutex);
+}
+EXPORT_SYMBOL_GPL(constframe_framedimensions_copy_prim);
+
 void constframe_constantcolor(struct dpu_constframe *cf, unsigned int r,
 			      unsigned int g, unsigned int b, unsigned int a)
 {
diff --git a/include/video/dpu.h b/include/video/dpu.h
index 7ac895d..c23a14e 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -506,6 +506,7 @@ struct dpu_fetchunit {
 void constframe_shden(struct dpu_constframe *cf, bool enable);
 void constframe_framedimensions(struct dpu_constframe *cf, unsigned int w,
 				unsigned int h);
+void constframe_framedimensions_copy_prim(struct dpu_constframe *cf);
 void constframe_constantcolor(struct dpu_constframe *cf, unsigned int r,
 			      unsigned int g, unsigned int b, unsigned int a);
 void constframe_controltrigger(struct dpu_constframe *cf, bool trigger);
-- 
1.7.9.5

